### PR TITLE
Scarlettnew

### PR DESCRIPTION
**Pull request by x811zou**
update code to allow BEASTIE to run on real data

### CODE REVIEW

File: BEASTIE/beastie_step2.py [hunk 1]
```diff
@@ -65,38 +62,40 @@ def run(hetSNP_intersect_unique,meta,hetSNP_intersect_unique_forlambda_file,hetS
...
     logging.info('..... finishing running binomial')
 
     ##########################
     logging.info('=================')
     logging.info('================= Starting step 2.5')
     logging.info('..... start generating gene list')
-    outfilename=out+"/"+prefix+"_ASE_all.tsv"
-    outfilename_ase=out+"/"+prefix+"_ASE_cutoff_"+str(cutoff)+"_filtered.tsv"
+    outfilename=result+"/"+prefix+"_ASE_all.tsv"
+    outfilename_ase=result+"/"+prefix+"_ASE_cutoff_"+str(cutoff)+"_filtered.tsv"
```
Comment (zackgomez) [inline BEASTIE/beastie_step2.py:87]: path.join

File: BEASTIE/predict_lambda_phasingError.R [hunk 2]
```diff
@@ -47,40 +47,41 @@ meta_error=args[9]
...
-# hetSNP_intersect_unique_forlambda_file="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/BEASTIE_example/HG00096_chr21/output/TEMP/HG00096_chr21_hetSNP_intersect_unique_forLambda.tsv"
+# hetSNP_intersect_unique_forlambda_file="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/other_example/HG00096/output/s-0.5_a-0.05_sinCov0_totCov1_W1000K1000/HG00096_hetSNP_intersected_filtered_forLambda.TEMP.tsv"
 # hetSNP_intersect_unique_lambdaPredicted_file="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/BEASTIE_example/HG00096_chr21/output/TEMP/HG00096_chr21_hetSNP_intersect_unique_lambdaPredicted.tsv"
 # meta="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/BEASTIE_example/HG00096_chr21/output/TEMP/HG00096_chr21_meta.tsv"
 # meta_error="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/BEASTIE_example/HG00096_chr21/output/HG00096_chr21_meta_w_error.tsv"
 
-source("Get_phasing_error_rate.R")
-source("Get_LD.R")
+source("./BEASTIE/Get_phasing_error_rate.R")
+source("./BEASTIE/Get_LD.R")
```
Comment (zackgomez) [inline BEASTIE/predict_lambda_phasingError.R:56]: resolve these merge conflicts by taking the master branch (current_changes)

File: BEASTIE/predict_lambda_phasingError.R [hunk 3]
```diff
@@ -47,40 +47,41 @@ meta_error=args[9]
...
 ############################################## 1. predict lambda #####################################################
 if(grepl("iBEASTIE", model, fixed=TRUE)){
-  lambda.fit.simulation<- readRDS("LinearReg_iBEASTIE_fitted_model_lambda.rds")
+  lambda.fit.simulation<- readRDS("./BEASTIE/LinearReg_iBEASTIE_fitted_model_lambda_loglambda_version2.rds")
 }else{
-  lambda.fit.simulation<- readRDS("LinearReg_BEASTIE_fitted_model_lambda.rds")
+  lambda.fit.simulation<- readRDS("./BEASTIE/LinearReg_BEASTIE_fitted_model_lambda_version2.rds")
 }
-
+lambda.fit.simulation<- readRDS("./BEASTIE/LinearReg_iBEASTIE_fitted_model_lambda_loglambda_version2.rds")
```
Comment (zackgomez) [inline BEASTIE/predict_lambda_phasingError.R:77]: Be sure to update this call to match the way it's done in master when merging

File: BEASTIE/annotate_LD_new.R [hunk 4]
```diff
@@ -13,10 +13,20 @@ infile=args[3]
 outpath=args[4]
 mytoken=args[5]
 chr_start=args[6]
-chr_end=args[7]
-meta=args[8]
+chr_end=args[7]  
+meta=args[8]     
+
+# prefix="NA19248"
```
Comment (tncowart) [inline BEASTIE/annotate_LD_new.R:19]: Please delete these comments

File: BEASTIE/__init__.py [hunk 5]
```diff
@@ -70,29 +69,53 @@ def load_configuration(config_file):
...
+        Path(result_path).mkdir(parents=True,exist_ok=True)
+
+    logname = os.path.join(config.input_dir,config.prefix,config.output_folder,specification,"log",f"{config.prefix}-{today.strftime('%b-%d-%Y')}.log")
     if os.path.isfile(logname):
         os.remove(logname)
     logging.basicConfig(filename=logname,
                             filemode='a',
                             format='%(asctime)-15s [%(levelname)s] %(message)s',
                             level=logging.DEBUG)
+    #stderr_file=os.path.join(config.input_dir,config.output_folder,config.specification,"/log/",config.prefix+"-"+str(today.strftime("%b-%d-%Y"))+".stderr")
```
Comment (tncowart) [inline BEASTIE/__init__.py:103]: if this code is no longer needed please delete this comment.

File: BEASTIE/annotation.py [hunk 6]
```diff
@@ -33,6 +34,7 @@ def annotateAF(ancestry, hetSNP, out_AF, ref_dir):
             AF=AF.rename(columns={'SAS_AF': 'AF'})
         data_AF=pd.merge(data,AF,on=['chr','pos'],how='left')
         data_AF=data_AF.drop_duplicates()
+        #out_AF="NA19248_hetSNP_chr1-22_AF.tsv"
```
Comment (tncowart) [inline BEASTIE/annotation.py:37]: If this is no longer needed, please delete it.

File: BEASTIE/Get_LD.R [hunk 7]
```diff
@@ -80,12 +80,15 @@ loop_chr <-function(outpath,data,ancestry,batch_size,mytoken,chr_start,chr_end){
...
       }else{
-        print(paste0(">>>>>>>>>> no content on chr",j," on hetSNP file!")) 
-        opt <- options(show.error.messages=TRUE)
-        on.exit(options(opt))
-        stop()
+        print(paste0("!!!!! only 1 site or no content on chr",j," on hetSNP file!")) 
+        counter=counter+1
+        ff=last_counter+1
+        last_counter=ff
+        #opt <- options(show.error.messages=TRUE)
```
Comment (tncowart) [inline BEASTIE/Get_LD.R:87]: if these comments are no longer needed, please delete them.

File: BEASTIE/annotate_LD_new.R [hunk 8]
```diff
@@ -4,7 +4,7 @@ suppressMessages(library(data.table))
 suppressMessages(library(foreach))
 suppressMessages(library("dplyr"))
 library("LDlinkR")
-source("Get_LD.R")
+source("./BEASTIE/Get_LD.R")
```
Comment (tncowart) [inline BEASTIE/annotate_LD_new.R:7]: You're going to have a merge conflict here, please resolve in favor of the other code.

File: BEASTIE/beastie_step1.py [hunk 9]
```diff
@@ -2,21 +2,18 @@
 #=========================================================================
 # 2021 Xue Zou (xue.zou@duke.edu)
 #=========================================================================
-import logging
 import os
 import sys
-
+import logging
```
Comment (tncowart) [inline BEASTIE/beastie_step1.py:7]: why did you move this? The imports should be in alphabetical order.

File: BEASTIE/beastie_step1.py [hunk 10]
```diff
@@ -2,21 +2,18 @@
...
 import pandas as pd
-
+from pathlib import Path
```
Comment (tncowart) [inline BEASTIE/beastie_step1.py:9]: this should be moved down a line -- all the "import"s should be together and all the "from"s should be together.

File: BEASTIE/beastie_step1.py [hunk 11]
```diff
@@ -159,7 +172,7 @@ def run(sigma,alpha,WARMUP,KEEPER,prefix,vcf_sample_name,in_path,out,model,vcf,r
 
    ##### 1.3 Annotation: AF
     hetSNP_AF = f"{os.path.splitext(hetSNP)[0]}_AF.tsv"
-    print(hetSNP_AF)
+    #print(hetSNP_AF)
```
Comment (tncowart) [inline BEASTIE/beastie_step1.py:175]: please delete this.

File: BEASTIE/beastie_step2.py [hunk 12]
```diff
@@ -65,38 +62,40 @@ def run(hetSNP_intersect_unique,meta,hetSNP_intersect_unique_forlambda_file,hetS
...
+    logging.info('..... start running iBEASTIE model')
+    df_ibeastie,picklename = run_model_stan_wrapper.run(specification,prefix,base_modelin_error,sigma,alpha,model,result,hetSNP_intersect_unique_lambdaPredicted_file,WARMUP,KEEPER,either_cov,total_cov)
+    if df_ibeastie.shape[0]>2:
         logging.info('..... model output is finished!')
     else:
         logging.error('..... model output is empty, please try again!')
         sys.exit(1)
 
-    df_adm = ADM_for_real_data.run(prefix,base_modelin,out,picklename)
+    #df_beastie,_ = run_model_stan_wrapper_beastie.run(prefix,base_modelin_error,sigma,alpha,"/Users/scarlett/allenlab/software/cmdstan/examples/BEASTIE/BEASTIE",result,hetSNP_intersect_unique_lambdaPredicted_file,WARMUP,KEEPER,either_cov,total_cov)
```
Comment (tncowart) [inline BEASTIE/beastie_step2.py:73]: if they are no longer needed, please delete these comments.

File: BEASTIE/binomial_for_real_data.py [hunk 13]
```diff
@@ -92,15 +92,15 @@ def getBaseline_pooled(fields,depth,hets):
         return (None,None,None,None)
 
 def run(prefix,inFile,out,picklename):
-    # prefix="HG00096_chr21"
-    # inFile="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/BEASTIE_example/HG00096_chr21/output/s-0.5_a-0.05_sinCov0_totCov1_W1000K1000/HG00096_chr21_hetSNP_intersected_filtered.TEMP.modelinput.tsv"
-    # out="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/BEASTIE_example/HG00096_chr21/output/s-0.5_a-0.05_sinCov0_totCov1_W1000K1000"
-    # picklename="HG00096_chr21_a-0.05_W1000K1000_s0t1.pickle"
+    # prefix="NA19257"
```
Comment (tncowart) [inline BEASTIE/binomial_for_real_data.py:95]: if these comments are no longer needed please delete them.

File: BEASTIE/binomial_for_real_data.py [hunk 14]
```diff
@@ -92,15 +92,15 @@ def getBaseline_pooled(fields,depth,hets):
...
+    # inFile="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/other_example/"+prefix+"/output/s-0.5_a-0.05_sinCov0_totCov1_W1000K1000/"+prefix+"_hetSNP_intersected_filtered.TEMP.modelinput.tsv"
+    # out="/Users/scarlett/Documents/Allen_lab/github/BEASTIE/other_example/"+prefix+"/output/s-0.5_a-0.05_sinCov0_totCov1_W1000K1000"
+    # picklename=prefix+"_a-0.05_W1000K1000_s0t1.pickle"
     outfix=picklename
     out_path=out+"/output_pkl/binomial"
-    # single site
-    out_path_1 = out_path+"/SS_esti/"
-    out_path_2 = out_path+"/SS_p/"
+    # First site
+    out_path_1 = out_path+"/FS_esti/"
```
Comment (tncowart) [inline BEASTIE/binomial_for_real_data.py:102]: os.path.join

File: BEASTIE/run_model_stan_wrapper.py [hunk 15]
```diff
@@ -154,86 +152,96 @@ def getCredibleInterval(thetas,alpha):
...
+    # if "txt" in inFile:
+    #     outfix=os.path.split(inFile)[1].rsplit("_hetSNP_intersected_filtered.TEMP.txt")[0]
+    # if "tsv" in inFile:
+    #     outfix=os.path.split(inFile)[1].rsplit("_hetSNP_intersected_filtered.TEMP.tsv")[0]
     tmpFile= "tmp_output.txt"
     initFile = "initialization_stan.txt"
     outFile= "stan_output.txt"
-    out=out+"/output_pkl"
-    out_path=out+"/beastie/"
+    out=out0+"/output_pkl"
```
Comment (tncowart) [inline BEASTIE/run_model_stan_wrapper.py:226]: os.path.join

File: BEASTIE/beastie_step2.py [hunk 16]
```diff
@@ -65,38 +62,40 @@ def run(hetSNP_intersect_unique,meta,hetSNP_intersect_unique_forlambda_file,hetS
...
         logging.info('..... output file save to {0}'.format(outfilename_ase))
-    significant_genes(df_beastie,df_binomial,df_adm,outfilename,outfilename_ase,cutoff,hetSNP_intersect_unique_lambdaPredicted_file)
-    if str(SAVE_INT) == "True":
-        for files in os.listdir(out):
-            if "TEMP" in files:
-                logging.info('..... remove TEMP files {0}'.format(files))
-                os.remove(out+"/"+files)
+    significant_genes(df_ibeastie,df_binomial,df_adm,outfilename,outfilename_ase,cutoff,hetSNP_intersect_unique_lambdaPredicted_file)
+    logging.info('..... done with significant_gene')
+    if str(SAVE_INT) == "False":
```
Comment (tncowart) [inline BEASTIE/beastie_step2.py:96]:
```suggestion
    if not SAVE_INT:
```

File: BEASTIE/BEASTIE.py [hunk 17]
```diff
@@ -59,7 +58,7 @@ def main():
     build_parser.add_argument('--alpha', type=float, default=0.05, help='Significance cutoff for ASE. Defaults to 0.5.')
     build_parser.add_argument('--sigma', type=float, default=0.5, help='Significance cutoff for ASE. Defaults to 0.5.')
     build_parser.add_argument('--out', default='output', help='Location to write output files. Defaults to \'output\' inside the sample folder')
-    build_parser.add_argument('--SAVE_INT', default='False', help='Whether to delte TEMP folder where it stores intermediate output. Defaults to keep \'TEMP\' folder inside \'output\' folder')
+    build_parser.add_argument('--SAVE_INT', default=False, help='Whether to delte TEMP folder where it stores intermediate output. Defaults to keep \'TEMP\' folder inside \'output\' folder')
```
Comment (tncowart) [inline BEASTIE/BEASTIE.py:61]:
```suggestion
    build_parser.add_argument('--SAVE_INT', help="Save the TEMP folder that stores intermediate output.", action="store_true")
```

**Review by tncowart — CHANGES_REQUESTED** [review]

**Comment by tncowart** [comment]
@x811zou You should almost always delete commented out code before committing it. If you do have a reason for leaving it, you should add a comment explaining why you left it. Remember that the code here is for other people to view and use.

File: BEASTIE/__init__.py [hunk 18]
```diff
@@ -70,29 +69,53 @@ def load_configuration(config_file):
...
     logging.info('======================================== ')
     hetSNP_intersect_unique, meta, hetSNP_intersect_unique_forlambda_file, hetSNP_intersect_unique_lambdaPredicted_file = beastie_step1.run(
-        0.5,  # sigma
-        config.alpha,
-        config.WARMUP,
-        config.KEEPER,
+        specification,
+        0.5,      # sigma
+        0.05,     # alpha
+        1000,     # WARMUP
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:117]: I will fix it

File: BEASTIE/__init__.py [hunk 19]
```diff
@@ -55,9 +53,10 @@ def load_configuration(config_file):
         ref_dir=inputs['ref_dir'],
         input_dir=inputs['input_dir'],
         SAVE_INT=inputs['SAVE_INT'],
-        WARMUP=inputs['WARMUP'],
-        KEEPER=inputs['KEEPER'],
-        output_dir=outputs['out_path'],
+        warmup=inputs['WARMUP'],
+        keeper=inputs['KEEPER'],
+        output_folder=outputs['folder'],
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:58]: Let me think about it

File: BEASTIE/__init__.py [hunk 20]
```diff
@@ -8,12 +8,12 @@
...
 import os.path
 from collections import namedtuple
+from pathlib import Path
 from datetime import date
-
 from . import beastie_step1, beastie_step2
 
 ConfigurationData = namedtuple("ConfigurationData", [
-    "prefix", "vcf_file_name", "vcf_sample_name", "pileup_file_name", "ancestry", "min_total_cov", "min_single_cov", "sigma", "cutoff", "alpha", "chr_start", "chr_end", "read_length", "LD_token", "modelName", "STAN", "work_dir", "ref_dir", "input_dir", "SAVE_INT", "WARMUP", "KEEPER", "output_dir"
+    "prefix", "vcf_file_name", "vcf_sample_name", "pileup_file_name", "ancestry", "min_total_cov", "min_single_cov", "sigma", "cutoff", "alpha", "chr_start", "chr_end", "read_length", "LD_token", "modelName", "STAN", "work_dir", "ref_dir", "input_dir", "SAVE_INT", "warmup", "keeper", "output_dir","output_folder"
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:16]: Just to keep them consistent with other parts of the code

[the remark this answers is not in the document]
Comment (tncowart) [reply BEASTIE/__init__.py:16]: I guess that's my point -- in some places you have `WARMUP` and `KEEPER` and now in other places you have `warmup` and `keeper`. I think it would be best to pick one or the other.

[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:16]: removed it

File: BEASTIE/__init__.py [hunk 21]
```diff
@@ -70,29 +69,53 @@ def load_configuration(config_file):
...
     model = os.path.join(f"{config.STAN}{config.modelName}", config.modelName)
     today = date.today()
 
-    logname = os.path.join(in_path, "output", f"{config.prefix}-{today.strftime('%b-%d-%Y')}.log")
+    specification="s"+str(config.sigma)+"_a"+str(config.alpha)+"_sinCov"+str(config.min_single_cov)+"_totCov"+str(config.min_total_cov)+"_W"+str(config.warmup)+"K"+str(config.keeper)
+    output_path=str(out_dir)+"/"+str(config.output_folder)
+    specification_path=output_path+"/"+str(specification)
+    log_path=specification_path+"/log/"
+    tmp_path=specification_path+"/tmp/"
+    result_path=specification_path+"/result/"
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:83]: changed it

File: BEASTIE/__init__.py [hunk 22]
```diff
@@ -70,29 +69,53 @@ def load_configuration(config_file):
...
+    if not os.path.exists(output_path):
+        Path(output_path).mkdir(parents=True,exist_ok=True)
+    if not os.path.exists(specification_path):
+        Path(specification_path).mkdir(parents=True,exist_ok=True)
+    if not os.path.exists(log_path):
+        Path(log_path).mkdir(parents=True,exist_ok=True)
+    if not os.path.exists(tmp_path):
+        Path(tmp_path).mkdir(parents=True,exist_ok=True)
+    if not os.path.exists(result_path):
+        Path(result_path).mkdir(parents=True,exist_ok=True)
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:94]: got it

File: BEASTIE/__init__.py [hunk 23]
```diff
@@ -70,29 +69,53 @@ def load_configuration(config_file):
...
     hetSNP_intersect_unique, meta, hetSNP_intersect_unique_forlambda_file, hetSNP_intersect_unique_lambdaPredicted_file = beastie_step1.run(
-        0.5,  # sigma
-        config.alpha,
-        config.WARMUP,
-        config.KEEPER,
+        specification,
+        0.5,      # sigma
+        0.05,     # alpha
+        1000,     # WARMUP
+        1000,     # KEEPER
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/__init__.py:118]: ok

File: BEASTIE/beastie_step1.py [hunk 24]
```diff
@@ -101,28 +98,44 @@ def check_file_existence(prefix,in_path,out,model,vcf,ref_dir,pileup,hetSNP,pars
...
+       logging.info('Creating COMMON directory : {0}'.format(common))
    else:
-       logging.info('Found existed output directory : {0}'.format(common))
-   out=common+"/s-"+str(sigma)+"_a-"+str(alpha)+"_sinCov"+str(min_single_cov)+"_totCov"+str(min_total_cov)+"_W"+str(WARMUP)+"K"+str(KEEPER)
-   if not os.path.isdir(out):
-      logging.info('Creating specific output directory : {0}'.format(out))
-      os.mkdir(out)
+       logging.info('Found existed COMMON directory : {0}'.format(common))
+
+   out=common+"/"+specification
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/beastie_step1.py:112]: ok

File: BEASTIE/beastie_step1.py [hunk 25]
```diff
@@ -101,28 +98,44 @@ def check_file_existence(prefix,in_path,out,model,vcf,ref_dir,pileup,hetSNP,pars
...
+      logging.info('Creating specific TEMP directory : {0}'.format(temp))
+      Path(temp).mkdir(parents=True,exist_ok=True)
    else:
-      #shutil.rmtree(out)
-      #os.mkdir(out)
-      logging.info('Found existed specific output directory : {0}, remove content ...'.format(out))
-   return out,common
+      logging.info('Found existed specific TEMP directory : {0}, remove content ...'.format(temp))
+
+   result = out+'/result'
```
[the remark this answers is not in the document]
Comment (x811zou) [reply BEASTIE/beastie_step1.py:127]: ok

File: BEASTIE/beastie_step2.py (hunk 1 above)
Comment (x811zou) [reply BEASTIE/beastie_step2.py:87]: ok
